### PR TITLE
Header discipline in base

### PR DESCRIPTION
**Pull request by dellaert**
Trying to cut down on compile times, so
- moved serialization of Vector and Matrix into separate headers, sparsely included
- moved various other includes out of base header files
- fixed issues related to those

### CODE REVIEW

File: gtsam/base/Matrix.cpp [hunk 1]
```diff
@@ -33,6 +34,7 @@
 #include <fstream>
 #include <limits>
 #include <iostream>
+#include <iostream>
```
Comment (varunagrawal) [inline gtsam/base/Matrix.cpp:37]: duplicate

Comment (dellaert) [reply gtsam/base/Matrix.cpp:37]: If you reviewed everything and you trust me to fix the above, approve??

**Review by varunagrawal — COMMENTED** [review]
There is this cool tool called [include-what-you-use](https://include-what-you-use.org/) which we should probably leverage.

File: gtsam_unstable/slam/tests/testSerialization.cpp [hunk 2]
```diff
@@ -18,12 +15,16 @@
...
-#include <sstream>
+#include <CppUnitLite/TestHarness.h>
+
 #include <boost/assign/std/vector.hpp>
 #include <boost/filesystem.hpp>
 
+#include <iostream>
+#include <cstdlib>
+#include <fstream>
+#include <sstream>
```
Comment (varunagrawal) [inline gtsam_unstable/slam/tests/testSerialization.cpp:26]: Pretty sure fstream includes iostream for us, but I like being explicit so yay!

**Review by varunagrawal — APPROVED** [review]
Took a deeper look and LGTM

**Comment by dellaert** [comment]
> There is this cool tool called [include-what-you-use](https://include-what-you-use.org/) which we should probably leverage.

Yeah, tried that at one point. Its not so great for large repos with boost and Eigen. Maybe we should try it again.

**Pull request merged**